File: php-checks/src/test/resources/checks/security/CryptographicHashCheck.php
<?php

$str = "foo";

hash($str);
foo($str);
sha1_file($str);
hash(algo: $str); 

  md5($str);     // Noncompliant {{Make sure this weak hash algorithm is not used in a sensitive context here.}}
//^^^
md5($str, true); // Noncompliant {{Make sure this weak hash algorithm is not used in a sensitive context here.}}
sha1($str);       // Noncompliant {{Make sure this weak hash algorithm is not used in a sensitive context here.}}
SHA1($str, false); // Noncompliant {{Make sure this weak hash algorithm is not used in a sensitive context here.}}

hash_init("md2"); // Noncompliant
//        ^^^^^
hash_init("md4"); // Noncompliant
hash_init("md5"); // Noncompliant
hash_init("sha1"); // Noncompliant
hash_init("sha224"); // Noncompliant
hash_init("ripemd128"); // Noncompliant
hash_init("ripemd160"); // Noncompliant
hash_init("haval160,3"); // Noncompliant
hash_init("haval192,3"); // Noncompliant
hash_init("haval224,3"); // Noncompliant
hash_init(1, algo: "haval224,3"); // Noncompliant
hash_init("sha512"); // Compliant
hash_init($unknown); // Compliant
hash_init(algo: $unknown); // Compliant

hash("md2", $data); // Noncompliant
hash("md4", $data); // Noncompliant
hash("md5", $data); // Noncompliant
hash("sha1", $data); // Noncompliant
hash("sha224", $data); // Noncompliant
hash("ripemd128", $data); // Noncompliant
hash("ripemd160", $data); // Noncompliant
hash("haval160,3", $data); // Noncompliant
hash("haval192,3", $data); // Noncompliant
hash("haval224,3", $data); // Noncompliant
hash($data, $value, "foo", algo: "md2"); // Noncompliant
hash("sha512", $data); // Compliant
hash($unknown, $data); // Compliant
hash(algo: "sha512"); 

hash_pbkdf2("md2", $password, $salt, $iterations); // Noncompliant
hash_pbkdf2("md4", $password, $salt, $iterations); // Noncompliant
hash_pbkdf2("md5", $password, $salt, $iterations); // Noncompliant
hash_pbkdf2("sha1", $password, $salt, $iterations); // Noncompliant
hash_pbkdf2("sha224", $password, $salt, $iterations); // Noncompliant
hash_pbkdf2("ripemd128", $password, $salt, $iterations); // Noncompliant
hash_pbkdf2("ripemd160", $password, $salt, $iterations); // Noncompliant
hash_pbkdf2("haval160,3", $password, $salt, $iterations); // Noncompliant
hash_pbkdf2("haval192,3", $password, $salt, $iterations); // Noncompliant
hash_pbkdf2("haval224,3", $password, $salt, $iterations); // Noncompliant
hash_pbkdf2($password, $salt, $iterations, algo: "haval224,3"); // Noncompliant
hash_pbkdf2("sha512", $password, $salt, $iterations); // Compliant
hash_pbkdf2($unknown, $password, $salt, $iterations); // Compliant
hash_pbkdf2($password, $salt, $iterations, algo: "sha512"); // Compliant

mhash(MHASH_MD2, $data); // Noncompliant
mhash(MHASH_MD4, $data); // Noncompliant
mhash(MHASH_MD5, $data); // Noncompliant
mhash(MHASH_RIPEMD128, $data); // Noncompliant
mhash(MHASH_SHA1, $data); // Noncompliant
mhash(MHASH_SHA192, $data); // Noncompliant
mhash(MHASH_SHA224, $data); // Noncompliant
mhash(MHASH_HAVAL128, $data); // Noncompliant
mhash(MHASH_HAVAL160, $data); // Noncompliant
mhash(MHASH_HAVAL192, $data); // Noncompliant
mhash(MHASH_HAVAL224, $data); // Noncompliant
mhash(MHASH_HAVAL224, $data); // Noncompliant
mhash($data, hash: MHASH_HAVAL224); // Noncompliant
mhash(MHASH_SHA512, $data); // Compliant
mhash("xyz", $data); // Compliant
mhash(algo: "xyz", $data); // Compliant

// Compliant: truncated output signals a non-cryptographic use (cache key, ETag, short identifier, etc.)
substr(md5($data), 0, 8);
substr(sha1($url), 0, 4);
substr(hash('sha1', $email), 0, 5);
substr(hash('sha1', $seed), -6);
mb_substr(md5($data), 0, 8);
mb_substr(sha1($data), 0, 4);
SUBSTR(md5($data), 0, 8);

// Hash used as length argument rather than string to truncate - still noncompliant
substr($str, 0, md5($str));  // Noncompliant {{Make sure this weak hash algorithm is not used in a sensitive context here.}}
//              ^^^
substr($str, 0, sha1($str)); // Noncompliant {{Make sure this weak hash algorithm is not used in a sensitive context here.}}
substr($str, 0, hash('sha1', $str)); // Noncompliant
//                   ^^^^^^

// Wrapped in something other than substr/mb_substr - still noncompliant
strtolower(md5($data)); // Noncompliant {{Make sure this weak hash algorithm is not used in a sensitive context here.}}
//         ^^^
strtolower(sha1($data)); // Noncompliant {{Make sure this weak hash algorithm is not used in a sensitive context here.}}

// Dynamic/variable callee - still noncompliant, no NPE on null function name
$fn(md5($data)); // Noncompliant {{Make sure this weak hash algorithm is not used in a sensitive context here.}}

// Compliant: hash used as a WordPress cache/transient key - not a security primitive
wp_cache_get(md5($sql), 'group');
wp_cache_set(sha1($sql), $value, 'group');
wp_cache_add(md5($sql), $value);
wp_cache_delete(md5($sql), 'group');
wp_cache_replace(md5($sql), $value);
wp_cache_incr(md5($sql));
wp_cache_decr(sha1($sql));
get_transient(md5($sql));
set_transient(sha1($query), $result, HOUR_IN_SECONDS);
set_transient('tx_' . sha1($query), $result, HOUR_IN_SECONDS);
set_transient('prefix_' . md5($query) . '_suffix', $result, HOUR_IN_SECONDS);
wp_cache_get('key_' . hash('sha1', $sql), 'group');
delete_transient(md5($sql));
wp_cache_get(hash('sha1', $sql), 'group');

// Named argument places hash in the value slot, not the key slot - still noncompliant
wp_cache_set(value: md5($value), key: $key); // Noncompliant {{Make sure this weak hash algorithm is not used in a sensitive context here.}}
//                  ^^^

// Hash not in the key position (not first argument) - still noncompliant
wp_cache_set($key, md5($value)); // Noncompliant {{Make sure this weak hash algorithm is not used in a sensitive context here.}}
//                 ^^^
wp_cache_set($key, sha1($value)); // Noncompliant {{Make sure this weak hash algorithm is not used in a sensitive context here.}}
wp_cache_set($key, hash('sha1', $value)); // Noncompliant
//                      ^^^^^^

// Hash passed to a non-cache function or dynamic callee - still noncompliant, no NPE on null function name
some_function(md5($data)); // Noncompliant {{Make sure this weak hash algorithm is not used in a sensitive context here.}}
//            ^^^

File: php-checks/src/main/java/org/sonar/php/checks/security/CryptographicHashCheck.java
/*
 * SonarQube PHP Plugin
 * Copyright (C) 2010-2026 SonarSource Sàrl
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the Sonar Source-Available License Version 1, as published by SonarSource Sàrl.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
package org.sonar.php.checks.security;

import java.util.Map;
import java.util.Set;
import org.sonar.check.Rule;
import org.sonar.php.checks.utils.CheckUtils;
import org.sonar.php.checks.utils.argumentmatching.ArgumentVerifierValueContainment;
import org.sonar.php.checks.utils.argumentmatching.FunctionArgumentCheck;
import org.sonar.plugins.php.api.tree.Tree;
import org.sonar.plugins.php.api.tree.expression.ExpressionTree;
import org.sonar.plugins.php.api.tree.expression.FunctionCallTree;

@Rule(key = "S4790")
public class CryptographicHashCheck extends FunctionArgumentCheck {

  private static final String MESSAGE = "Make sure this weak hash algorithm is not used in a sensitive context here.";

  private static final Set<String> WEAK_HASH_FUNCTIONS = Set.of("md5", "sha1");
  private static final Set<String> SUBSTR_FUNCTIONS = Set.of("substr", "mb_substr");
  // Maps each WordPress cache/transient function to its key parameter name, used to resolve
  // the key argument correctly even when PHP 8 named arguments reorder the call.
  private static final Map<String, String> WORDPRESS_CACHE_KEY_PARAMS = Map.of(
    "wp_cache_get", "key",
    "wp_cache_set", "key",
    "wp_cache_add", "key",
    "wp_cache_delete", "key",
    "wp_cache_replace", "key",
    "wp_cache_incr", "key",
    "wp_cache_decr", "key",
    "get_transient", "transient",
    "set_transient", "transient",
    "delete_transient", "transient");
  private static final Set<String> WEAK_HASH_ARGUMENTS = Set.of(
    "md2",
    "md4",
    "md5",
    "sha1",
    "sha224",
    "ripemd128",
    "ripemd160",
    "haval160,3",
    "haval192,3",
    "haval224,3");
  private static final Set<String> WEAK_MHASH_ARGUMENTS = Set.of(
    "MHASH_MD2",
    "MHASH_MD4",
    "MHASH_MD5",
    "MHASH_RIPEMD128",
    "MHASH_SHA1",
    "MHASH_SHA192",
    "MHASH_SHA224",
    "MHASH_HAVAL128",
    "MHASH_HAVAL160",
    "MHASH_HAVAL192",
    "MHASH_HAVAL224");

  private static final ArgumentVerifierValueContainment hashArgumentVerifier = ArgumentVerifierValueContainment.builder()
    .position(0)
    .name("algo")
    .values(WEAK_HASH_ARGUMENTS)
    .build();
  private static final ArgumentVerifierValueContainment mHashArgumentVerifier = ArgumentVerifierValueContainment.builder()
    .position(0)
    .name("hash")
    .values(WEAK_MHASH_ARGUMENTS)
    .build();

  @Override
  public void visitFunctionCall(FunctionCallTree tree) {
    super.visitFunctionCall(tree);

    String functionName = CheckUtils.getLowerCaseFunctionName(tree);
    if (functionName != null && WEAK_HASH_FUNCTIONS.contains(functionName)) {
      if (!isWrappedInSubstr(tree) && !isUsedAsWordPressCacheKey(tree)) {
        createIssue(tree);
      }
      return;
    }

    checkArgument(tree, "hash_init", hashArgumentVerifier);
    if (!isWrappedInSubstr(tree) && !isUsedAsWordPressCacheKey(tree)) {
      checkArgument(tree, "hash", hashArgumentVerifier);
    }
    checkArgument(tree, "hash_pbkdf2", hashArgumentVerifier);
    checkArgument(tree, "mhash", mHashArgumentVerifier);
  }

  // SONARPHP-1840: Suppress when the hash output is immediately truncated by substr/mb_substr,
  // which signals a non-cryptographic use (cache key, ETag, short identifier, etc.)
  private static boolean isWrappedInSubstr(FunctionCallTree hashCall) {
    Tree parent = hashCall.getParent();
    if (parent == null || !parent.is(Tree.Kind.CALL_ARGUMENT)) {
      return false;
    }
    Tree grandParent = parent.getParent();
    if (grandParent == null || !grandParent.is(Tree.Kind.FUNCTION_CALL)) {
      return false;
    }
    FunctionCallTree outerCall = (FunctionCallTree) grandParent;
    String outerName = CheckUtils.getLowerCaseFunctionName(outerCall);
    if (outerName == null || !SUBSTR_FUNCTIONS.contains(outerName)) {
      return false;
    }
    return !outerCall.callArguments().isEmpty() && outerCall.callArguments().get(0) == parent;
  }

  // SONARPHP-1839: Suppress when the hash output is used as a WordPress cache/transient key,
  // either directly or as part of a concatenated key string (e.g. 'prefix_' . sha1($q)).
  // Cache keys are not a security primitive — a collision yields a stale read, not a security incident.
  private static boolean isUsedAsWordPressCacheKey(FunctionCallTree hashCall) {
    Tree node = hashCall.getParent();
    while (node != null && node.is(Tree.Kind.CONCATENATION)) {
      node = node.getParent();
    }
    Tree callArg = node;
    if (callArg == null || !callArg.is(Tree.Kind.CALL_ARGUMENT)) {
      return false;
    }
    Tree grandParent = callArg.getParent();
    if (grandParent == null || !grandParent.is(Tree.Kind.FUNCTION_CALL)) {
      return false;
    }
    FunctionCallTree outerCall = (FunctionCallTree) grandParent;
    String outerName = CheckUtils.getLowerCaseFunctionName(outerCall);
    String keyParamName = outerName != null ? WORDPRESS_CACHE_KEY_PARAMS.get(outerName) : null;
    if (keyParamName == null) {
      return false;
    }
    return CheckUtils.argument(outerCall, keyParamName, 0)
      .filter(keyArg -> keyArg == callArg)
      .isPresent();
  }

  protected void createIssue(FunctionCallTree tree) {
    context().newIssue(this, tree.callee(), MESSAGE);
  }

  @Override
  protected void createIssue(ExpressionTree argument) {
    context().newIssue(this, argument, MESSAGE);
  }
}
